lib/logstash/sized_queue_timeout.rb

```diff
@@ -0,0 +1,59 @@
+require "concurrent/atomic/condition"
+require "thread"
+
+module LogStash
+  # Minimal subset implement of a SizedQueue supporting
+  # a timeout option on the lock.
+  #
+  # This will be part of the main Logstash's sized queue
+  class SizedQueueTimeout
+    class TimeoutError < StandardError; end
+
+    DEFAULT_TIMEOUT = 2 # in seconds
+
+    def initialize(max_size, options = {})
+      @condition_in = Concurrent::Condition.new
+      @condition_out = Concurrent::Condition.new
+
+      @max_size = max_size
+      @queue = []
+      @mutex = Mutex.new
+      @timeout = options.fetch(:timeout, DEFAULT_TIMEOUT)
+    end
+
+    def push(obj)
```

Comment:
+ 1 once we get the logic merge in the @colinsurprenant's persistent queue plugin developers could use their own tolerance values.